hazelcast/src/test/java/com/hazelcast/cp/CPGroupAvailabilityListenerTest.java

```diff
@@ -135,4 +136,80 @@ public void majorityLost(CPGroupAvailabilityEvent event) {
             majorityLatch.countDown();
         }
     }
+
+    static class GracefulShutdownAvailabilityListener
+            implements CPGroupAvailabilityListener {
+        private final AtomicInteger handlerHits;
+
+        GracefulShutdownAvailabilityListener() {
+            handlerHits = new AtomicInteger();
+        }
+
+        @Override
+        public void availabilityDecreased(CPGroupAvailabilityEvent event) {
+            if (event instanceof CPGroupAvailabilityEventGracefulImpl) {
+                handlerHits.incrementAndGet();
+            }
+        }
+
+        // majority is completely broken in this scenario
+        @Override
+        public void majorityLost(CPGroupAvailabilityEvent event) {
+            if (event instanceof CPGroupAvailabilityEventGracefulImpl) {
+                handlerHits.incrementAndGet();
+            }
+        }
+    }
+
+    @Test
+    public void whenMemberShutdown_thenReceiveEvents3Graceful() {
+        GracefulShutdownAvailabilityListener listener = new GracefulShutdownAvailabilityListener();
+
+        HazelcastInstance[] instances = newInstances(3);
+        HazelcastInstance member1 = instances[0];
+        HazelcastInstance member2 = instances[1];
+        HazelcastInstance member3 = instances[2];
+        member2.getCPSubsystem().addGroupAvailabilityListener(listener);
+        member3.getCPSubsystem().addGroupAvailabilityListener(listener);
+
+        int expectedAvailabilityEvents = 2;
+        member1.getLifecycleService().shutdown();
+        assertEqualsEventually(expectedAvailabilityEvents, listener.handlerHits);
+
+        expectedAvailabilityEvents += 1;
+        member2.getLifecycleService().shutdown();
+        // majority lost ---
+        assertEqualsEventually(expectedAvailabilityEvents, listener.handlerHits);
+    }
+
+    @Test
+    public void whenMemberShutdown_thenReceiveEvents5Graceful() {
+        GracefulShutdownAvailabilityListener listener = new GracefulShutdownAvailabilityListener();
+
+        HazelcastInstance[] instances = newInstances(5);
+        instances[1].getCPSubsystem().addGroupAvailabilityListener(listener);
+        instances[2].getCPSubsystem().addGroupAvailabilityListener(listener);
```

Comment:
I assume that there is no need to register listeners for every member.

Reply:
There's no need -- just seeing semantics of pub/sub on the events.

Reply:
Updated to just use one member to host the listener -- as I mentioned -- I was playing to see what the pub/sub semantics appeared to be.